sqlstore/src/test/java/uk/ac/bbsrc/tgac/miso/persistence/impl/HibernateProgressDaoTest.java

```diff
@@ -0,0 +1,243 @@
+
+import static org.junit.Assert.*;
+import static uk.ac.bbsrc.tgac.miso.core.data.workflow.Workflow.WorkflowType.LOADSEQUENCER;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Date;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.hibernate.SessionFactory;
+import org.junit.Before;
+import org.junit.Test;
+import org.mockito.InjectMocks;
+import org.mockito.MockitoAnnotations;
+import org.springframework.beans.factory.annotation.Autowired;
+
+import com.eaglegenomics.simlims.core.User;
+import com.google.common.collect.Streams;
+
+import uk.ac.bbsrc.tgac.miso.AbstractDAOTest;
+import uk.ac.bbsrc.tgac.miso.core.data.workflow.AbstractProgressStep.ProgressStepId;
+import uk.ac.bbsrc.tgac.miso.core.data.workflow.Progress;
+import uk.ac.bbsrc.tgac.miso.core.data.workflow.ProgressStep;
+import uk.ac.bbsrc.tgac.miso.core.data.workflow.Workflow.WorkflowType;
+import uk.ac.bbsrc.tgac.miso.core.data.workflow.impl.PoolProgressStep;
+import uk.ac.bbsrc.tgac.miso.core.data.workflow.impl.ProgressImpl;
+import uk.ac.bbsrc.tgac.miso.core.data.workflow.impl.SampleProgressStep;
+
+public class HibernateProgressDaoTest extends AbstractDAOTest {
+  private static final long UNKNOWN_WORKFLOW_PROGRESS_ID = 42;
+  private static final long WORKFLOW_PROGRESS_ID_1 = 1;
+  private static final long WORKFLOW_PROGRESS_ID_2 = 2;
+
+  private static final long UNKNOWN_USER_ID = 42;
+  private static final long USER_ID = 3;
+
+  // Progress objects that mirror most of the attributes of the objects in the test database
+  private Progress progress1;
+  private Progress progress2;
+
+  @InjectMocks
+  private HibernateProgressDao dao;
+
+  @Autowired
+  private SessionFactory sessionFactory;
+
+  @Before
+  public void setup() {
+    MockitoAnnotations.initMocks(this);
+    dao.setSessionFactory(sessionFactory);
+    progress1 = createProgress1();
+    progress2 = createProgress2();
+  }
+
+  @Test
+  public void getByUnknownId() {
+    assertNull(dao.get(UNKNOWN_WORKFLOW_PROGRESS_ID));
+  }
+
+  @Test
+  public void getByKnownId() {
+    assertEquivalent(progress1, dao.get(WORKFLOW_PROGRESS_ID_1));
+  }
+
+  @Test
+  public void listByUnknownUser() {
+    assertEquals(Collections.emptyList(), dao.listByUserId(UNKNOWN_USER_ID));
+  }
+
+  @Test
+  public void listByKnownUser() {
+    List<Progress> progresses = dao.listByUserId(USER_ID);
+
+    assertEquivalent(progress1, progresses.get(0));
+    assertEquivalent(progress2, progresses.get(1));
+  }
+
+  @Test
+  public void createWithOrderedSteps() {
+    Progress actual = dao.create(createProgressWithOrderedSteps());
+
+    assertEquivalent(createProgressWithOrderedSteps(), dao.get(actual.getId()));
+  }
+
+  @Test
+  public void updateNonExisting() {
+    Progress actual = createProgressWithOrderedSteps();
+
+    // When update is called on a non-existent entity, the entity is added to the database
```

Comment:
Didn't know that... Might as well combine save/update into one method like in some of the other DAOs

Reply:
Unintentionally, `ProgressStore#create` and `ProgressStore#update` have the same behaviour.  Ergo, I combined both methods into a single method `Progress save(Progress)`.